### DISCUSSION TITLE
Architecture advice on how to implement multiple actions on single route (RPC interface vs multiple API routes)

### DISCUSSION THREAD

**Discussion by magnusdr** [remix-run/remix#3898] [Q&A] [answered]
Let's say you have a route where you can do multiple actions. E.g. a shopping cart page: 
* remove product from cart, 
* change quantity,
* select shipping method etc. 

Do you think I should implement some RPC interface on the single `/cart`-route action handler? Or create separate API routes for every single action (`/api/remove-product-from-cart`, `/api/change-cart-qty`, `/api/select-shipping-method` etc)?

The latter seems to introduce a lot of bloat, and makes navigating the code base harder (connecting a page to the correct API routes). A RPC implementation however, can also quickly become over engineered. 

I'm considering using GraphQL, but this is only for data mutation, not querying, so a lot of the GraphQL features will be left unused and feels a little exaggerated for this use case. I'm already using GraphQL on the server side as explained in [this blog post](https://layercode.com/community/remix-apollo-graphql), and it feels nice not to bundle apollo with the client code. It would also be nice to "group" the mutations per route (`/cart`, `/profile`, `/product/:id`) instead of having a single `/graphql` API route for all mutations.

Any ideas? How are you solving this particular issue?

**Comment by sergiodxa** [comment, marked as answer]
I think you want something like I describe here https://sergiodxa.com/articles/multiple-forms-per-route-in-remix#the-action-reducer-pattern.

Pass a value in the form data that lets you identify what you want to do, then add a switch case or multiple ifs inside your action to act accordingly.

And as you say another option is to use resource routes for each action.

I personally like more a the action reducer pattern since it's simpler to open a route file and have all the actions the users can do there.

**Reply by edge-rps** [reply]
Video of this ⬆ https://www.youtube.com/watch?v=w2i-9cYxSdc&list=PLXoynULbYuEDG2wBFSZ66b85EIspy3fy6

**Reply by magnusdr** [reply]
Nice! This seems like a nice way of solving it. Great article @sergiodxa, and thanks for the video link @ryker2000  🙌
